src/commandsAndMenu.tsx

```diff
@@ -1620,13 +1620,28 @@ export async function showGitOperationDialog<T>(
         errorMessage => (error as Error).message.indexOf(errorMessage) > -1
       )
     ) {
+      const remoteURI = await model.getRemotes().then(remote => {
+        return remote;
+      });
```

Comment:
Good this is the idea indeed. But now we need to be smart depending on the `operation` type. If the operation is `Operation.Clone`, we don't need to request the remotes as we can look directly at `args.url`. In the cases of push and pull, we should not take `remoteURI[0]` but we should find the item matching `args.remote`. Then for the fetch case, let's use your fallback of only testing `remoteURI[0]`.

Reply:
@fcollonval  

**For Pull:**
Pull - I tried accessing args. When logging it, I noticed args is an empty object (no remote property), so I passed it in using the code below, which showed me that args.remote = 'origin'. But I don't think I fully understand what you mean by 'finding the item matching args.remote', do you mind explaining further? I guess I just don't see how the args.remote is going to tell us that the git provider is github (if that was what you were suggesting).

commandsAndMenu.tsx Line 383
```
  /** Add git pull command */
  commands.addCommand(CommandIDs.gitPull, {
   
    execute: async args => {
      try {
        // copy how commandIDS.gitPush grabs remote
        let remote;
        const result = await showDialog({
          title: trans.__('Please select push options.'),
          body: new AdvancedPushForm(trans, gitModel),
          buttons: [
            Dialog.cancelButton({ label: trans.__('Cancel') }),
            Dialog.okButton({ label: trans.__('Proceed') })
          ]
        });
        remote = result.value.remoteName;

        // added args as an additional param
        const details = await showGitOperationDialog(
          gitModel,
          Operation.Pull,
          trans,
          (args = { remote })
        );

```

**For Push**
- By default, args is an object with remote property, but it was undefined. Like for the pull, I'm not sure if I'm meant to pass in the remote by modifying

Reply:
`git` uses name for remote to be more user friendly. Then it uses a db to get the URI from the name.

You can see that by executing in a terminal `git remote -vv`; the output will look like:

```sh
origin  git@github.com:fcollonval/jupyterlab.git (fetch)
origin  git@github.com:fcollonval/jupyterlab.git (push)
trungleduc      https://github.com/trungleduc/jupyterlab (fetch)
trungleduc      https://github.com/trungleduc/jupyterlab (push)
upstream        https://github.com/jupyterlab/jupyterlab.git (fetch)
upstream        https://github.com/jupyterlab/jupyterlab.git (push)
```

Name `origin` refers `git@github.com:fcollonval/jupyterlab.git` in the above example.

---

As you discover (:clap: for that), you could get the name passing the `remote` variable. But to determine the associated URI, you first need to call `getRemote` (as you have done already). It returns you an array of `Git.IGitRemote`. That interface has the following definition:

```typescript
interface IGitRemote {
    url: string;
    name: string;
  }
```

So now that you know `remote` you need to find the `IGitRemote` object in the array with that name. Then you will be able to test the URL.

Does it clarify what you need to do?

---

Regarding push, if it is undefined, you could still apply the same logic and fallback to `remoteURI[0].url` if it is undefined.

> We could improve this by having the backend tell use which of the remote is the default. But let's try not to increase the scope of this PR too much.

Reply:
Hi @fcollonval !

Yes this was a very thorough explanation, thank you!

I was able to add the edits you requested for pull, push, clone, and fetch.

I have two pending issues/comments:
1. How do I test git fetch?
- I tried using `git fetch -all` which returned 'Fetching origin'
- However, it didn't throw any errors even though we are not authenticated, so I'm not sure how to proceed.
<img width="879" alt="image" src="https://user-images.githubusercontent.com/94336773/219998757-ea5651df-edc3-4338-96bc-aa0f151e8240.png">

2. In git pull, I passed the args to the git-pull catch block by using the following code. It's wrong since it creates an unnecessary dialog box just to get the remote URI. Is there some other way of getting the current remote name that is better? If not, I'll just remove this section as there are fallbacks in place already.
```
        let remote;
        const result = await showDialog({
          title: trans.__('Please select push options.'),
          body: new AdvancedPushForm(trans, gitModel),
          buttons: [
            Dialog.cancelButton({ label: trans.__('Cancel') }),
            Dialog.okButton({ label: trans.__('Proceed') })
          ]
        });
        remote = result.value.remoteName;
```

3. I grouped git pull, push and force in the catch block because that made sense to me, but there may be some nuances here that I'm overlooking.